File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.cpp
class Solution {
public:
    int f(string &s,int i,int sign,long ans){
        if(i>=s.size() || !isdigit(s[i])) return ans*sign;
        ans=ans*10+(s[i]-'0');
        if(sign*ans>INT_MAX) return INT_MAX;
        if(sign*ans<INT_MIN) return INT_MIN;

        return f(s,i+1,sign,ans);

    }
    int myAtoi(string s) {
        long ans = 0;
        int sign = 1;
        int i = 0;
        while (i < s.length() && s[i] == ' ') i++;
        if (i < s.length() && (s[i] == '-' || s[i] == '+')) {
            sign = (s[i] == '-') ? -1 : 1;
            i++;
        }
        return f(s,i,sign,ans);
    }
};

File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.java
class Solution {
    public int myAtoi(String s) {
        int ans=0;
        int n=s.length();

        int i=0;
        while(i<n && s.charAt(i)==' ') i++;

        int sign=1;
        if (i < n && (s.charAt(i) == '+' || s.charAt(i) == '-')) {
            if (s.charAt(i) == '-') sign = -1;
            i++;
        }

        

        while(i<n && Character.isDigit(s.charAt(i))){
            int dig=s.charAt(i)-'0';

            if (ans > (Integer.MAX_VALUE - dig) / 10) {
                return sign == 1 ? Integer.MAX_VALUE : Integer.MIN_VALUE;
            }
            
            ans = ans * 10 + dig;
            i++;
        }

        return ans*sign;
    }
}